firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SessionLifecycleClient.kt

```diff
@@ -0,0 +1,181 @@
+      )
+      appContext.bindService(
+        intent,
+        serviceConnection,
+        Context.BIND_IMPORTANT or Context.BIND_AUTO_CREATE
+      )
+    }
+  }
+
+  /**
+   * Should be called when any activity in this application process goes to the foreground. This
+   * will relay the event to the [SessionLifecycleService] where it can make the determination of
+   * whether or not this foregrounding event should result in a new session being generated.
+   */
+  fun foregrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.FOREGROUNDED)
+  }
+
+  /**
+   * Should be called when any activity in this application process goes from the foreground to the
+   * background. This will relay the event to the [SessionLifecycleService] where it will be used to
+   * determine when a new session should be generated.
+   */
+  fun backgrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.BACKGROUNDED)
+  }
+
+  /**
+   * Sends a message to the [SessionLifecycleService] with the given event code. This will
+   * potentially also send any messages that have been queued up but not successfully delivered to
+   * thes service since the previous send.
+   */
+  private fun sendLifecycleEvent(messageCode: Int) {
+    val allMessages = ArrayList<Message>()
+    queuedMessages.drainTo(allMessages)
+    allMessages.add(Message.obtain(null, messageCode, 0, 0))
+    sendLifecycleEvents(allMessages)
+  }
+
+  /**
+   * Sends lifecycle events to the [SessionLifecycleService]. This will only send the latest
+   * FOREGROUND and BACKGROUND events to the service that are included in the given list. Running
+   * through the full backlog of messages is not useful since the service only cares about the
+   * current state and transitions from background -> foreground.
+   */
+  private fun sendLifecycleEvents(messages: List<Message>) {
+    val latest = ArrayList<Message>()
+    getLatestByCode(messages, SessionLifecycleService.BACKGROUNDED)?.let { latest.add(it) }
+    getLatestByCode(messages, SessionLifecycleService.FOREGROUNDED)?.let { latest.add(it) }
+    latest.sortBy { it.getWhen() }
+
+    latest.forEach { sendMessageToServer(it) }
+  }
+
+  /** Sends the given [Message] to the [SessionLifecycleService]. */
+  private fun sendMessageToServer(msg: Message) {
+    if (service != null) {
+      try {
+        Log.i(TAG, "Sending lifecycle ${msg.what} at time ${msg.getWhen()} to service")
+        service?.send(msg)
+      } catch (e: RemoteException) {
+        Log.e(TAG, "Unable to deliver message: ${msg.what}")
+        queueMessage(msg)
+      }
+    } else {
+      queueMessage(msg)
+    }
+  }
+
+  /**
+   * Queues the given [Message] up for delivery to the [SessionLifecycleService] once the connection
+   * is established.
+   */
+  private fun queueMessage(msg: Message) {
+    Log.i(TAG, "Queueing message ${msg.what}")
+    while (queuedMessages.size >= MAX_QUEUED_MESSAGES) {
```

Comment:
This isn't really thread-safe (two threads could be doing this at the same time with unpredictable results). Though it's probably okay, we can also probably live without it.

Reply:
Removed.